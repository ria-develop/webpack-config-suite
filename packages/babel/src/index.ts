export * from './JsxExtensions';
export * from './HotModuleReplacementConfig';
export * from './BabelReactConfig';
export * from './BabelConfig';
