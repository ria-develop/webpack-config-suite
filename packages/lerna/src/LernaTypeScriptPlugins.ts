import {TypeScriptPlugins} from '@webpack-config-suite/typescript';
import {ForkTsCheckerWebpackPluginOptions} from 'fork-ts-checker-webpack-plugin/lib/ForkTsCheckerWebpackPluginOptions';

export class LernaTypeScriptPlugins extends TypeScriptPlugins {
  protected get typescript(): ForkTsCheckerWebpackPluginOptions['typescript'] {
    if (typeof super.typescript === 'object') {
      super.typescript.configFile = 'packages/tsconfig.json';
    }
    return super.typescript;
  }

  protected get eslint(): ForkTsCheckerWebpackPluginOptions['eslint'] {
    return {
      files: './packages/**/src/*.{ts,tsx,js,jsx}',
      enabled: true
    };
  }
}
