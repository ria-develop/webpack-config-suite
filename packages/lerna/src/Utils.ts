import * as path from 'path';

export const getContext = (): string => {
  return process.cwd();
};

export const resolve = (...pathSegments: string[]): string => {
  return path.resolve(...pathSegments);
};

export const findRootTsConfig = (): string => {
  try {
    return require.resolve(path.resolve(getContext(), 'packages/tsconfig.json'));
  } catch (e) {
    return 'tsconfig.json';
  }
};
