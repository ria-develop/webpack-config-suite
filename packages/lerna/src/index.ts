export * from './LernaTypeScriptResolvePlugins';
export * from './LernaTypeScriptPlugins';
export * from './LernaTypeScriptReactConfig';
