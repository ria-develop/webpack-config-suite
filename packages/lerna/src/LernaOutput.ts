import {BaseConfig, WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {getContext} from './Utils';

export class LernaOutput extends BaseConfig {
  protected get filename(): Configuration['output']['filename'] {
    return `./packages/[name]/lib/index${(this.isProd && '.min') || ''}.js`;
  }

  constructor(config: Configuration = {}, env: unknown = undefined, argv: WebpackCliOptions = {}) {
    super(config, env, argv);
    // TODO: core::Output.ts
    config.output = {
      path: getContext(),
      filename: this.filename
    };
  }
}
