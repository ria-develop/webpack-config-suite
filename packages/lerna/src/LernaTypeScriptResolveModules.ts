import {ResolveModules} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {getContext, resolve} from './Utils';

export class LernaTypeScriptResolveModules extends ResolveModules {
  protected get modules(): Configuration['resolve']['modules'] {
    return [resolve(getContext()), resolve(getContext(), './packages'), 'node_modules']; // TODO: get packages from @lerna/project
  }
}
