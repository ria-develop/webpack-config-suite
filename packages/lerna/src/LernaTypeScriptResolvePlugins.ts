import {ResolvePlugins} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {TsconfigPathsPlugin} from 'tsconfig-paths-webpack-plugin';
import {Options} from 'tsconfig-paths-webpack-plugin/lib/options';
import {findRootTsConfig} from './Utils';

export class LernaTypeScriptResolvePlugins extends ResolvePlugins {
  protected get configFile(): string {
    return findRootTsConfig();
  }

  protected get tsConfigPathsOptions(): Partial<Options> {
    return {configFile: this.configFile};
  }

  protected get plugins(): Configuration['resolve']['plugins'] {
    return [new TsconfigPathsPlugin(this.tsConfigPathsOptions)];
  }
}
