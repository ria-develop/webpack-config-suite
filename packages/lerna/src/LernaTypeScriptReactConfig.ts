import {TsExtension, TsLoaderOptions, TsxExtension, TypeScriptReactConfig} from '@webpack-config-suite/typescript';
import {JsExtension, What} from '@webpack-config-suite/core';
import {LernaTypeScriptResolvePlugins} from './LernaTypeScriptResolvePlugins';
import {HotModuleReplacementConfig} from '@webpack-config-suite/babel';
import {LernaTypeScriptPlugins} from './LernaTypeScriptPlugins';
import {LernaTypeScriptResolveModules} from './LernaTypeScriptResolveModules';
import {LernaOutput} from './LernaOutput';

export class LernaTypeScriptReactConfig extends TypeScriptReactConfig {
  protected get useThread(): boolean {
    return false;
  }

  public get composed(): What[] {
    return [
      LernaTypeScriptResolvePlugins,
      LernaTypeScriptResolveModules,
      JsExtension,
      TsExtension,
      TsxExtension,
      HotModuleReplacementConfig,
      LernaTypeScriptPlugins,
      LernaOutput
    ].filter(Boolean);
  }

  protected get tsLoaderOptions(): TsLoaderOptions {
    return {...super.tsLoaderOptions, projectReferences: true};
  }
}
