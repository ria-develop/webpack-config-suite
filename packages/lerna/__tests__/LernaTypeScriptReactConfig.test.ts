jest.mock('path', () => ({
  ...jest.requireActual('path'),
  resolve: jest.fn().mockImplementation((...pathSegments: string[]): string => {
    const resolved = jest.requireActual('path').resolve(...pathSegments);
    if (resolved.includes('src')) {
      return './src' + resolved.split('src')[1].replace(/\\/g, '/');
    }
    return resolved;
  })
}));
jest.mock('declaration-bundler-webpack-plugin');
jest.mock('tsconfig-paths-webpack-plugin');
jest.mock('../src/Utils', () => ({
  ...jest.requireActual('../src/Utils'),
  getContext: () => './',
  resolve: (...pathSegments: string[]) => pathSegments.join('/')
}));
import {WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {LernaTypeScriptReactConfig} from '../src';
import {TestCases, testProcessConfig} from '../../../jest';

describe.each(TestCases)(
  'given mode: "%s", env: "%j", argv: "%j"',
  (mode: Configuration['mode'], env: unknown, argv: WebpackCliOptions): void => {
    it('given TypeScriptConfig should match snapshot', () => {
      const config: Configuration = {mode};

      class MyLernaTypeScriptReactConfig extends LernaTypeScriptReactConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }

      testProcessConfig(MyLernaTypeScriptReactConfig, config, env, argv);
      expect(config).toMatchSnapshot();
    });
  }
);
