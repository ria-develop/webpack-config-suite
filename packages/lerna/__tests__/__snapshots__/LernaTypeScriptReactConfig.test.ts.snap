// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":false,"hot":false}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": false,
              "projectReferences": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "./packages/[name]/lib/index.js",
    "path": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./packages/**/src/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".ts",
      ".tsx",
    ],
    "modules": Array [
      "./",
      ".//./packages",
      "node_modules",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":true,"hot":true}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "plugins": Array [
                "react-hot-loader/babel",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": false,
              "projectReferences": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "./packages/[name]/lib/index.js",
    "path": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./packages/**/src/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react-dom": "@hot-loader/react-dom",
    },
    "extensions": Array [
      ".js",
      ".ts",
      ".tsx",
    ],
    "modules": Array [
      "./",
      ".//./packages",
      "node_modules",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
  "target": "web",
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"mode":"development"}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": false,
              "projectReferences": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "./packages/[name]/lib/index.js",
    "path": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./packages/**/src/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".ts",
      ".tsx",
    ],
    "modules": Array [
      "./",
      ".//./packages",
      "node_modules",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`given mode: "production", env: "undefined", argv: "{"liveReload":true,"hot":true}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": false,
              "projectReferences": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "./packages/[name]/lib/index.min.js",
    "path": "./",
  },
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".ts",
      ".tsx",
    ],
    "modules": Array [
      "./",
      ".//./packages",
      "node_modules",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;

exports[`given mode: "production", env: "undefined", argv: "{"mode":"production"}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": false,
              "projectReferences": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "./packages/[name]/lib/index.min.js",
    "path": "./",
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./packages/**/src/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".ts",
      ".tsx",
    ],
    "modules": Array [
      "./",
      ".//./packages",
      "node_modules",
    ],
    "plugins": Array [
      TsconfigPathsPlugin {
        "apply": [MockFunction],
      },
    ],
  },
}
`;
